circuit/src/gadgets/range.rs

```diff
@@ -0,0 +1,469 @@
+use crate::{
+	utils::{fe_to_le_bits, le_bits_to_u64},
+	RegionCtx,
+};
+use halo2::{
+	circuit::{AssignedCell, Layouter, Region, Value},
+	halo2curves::FieldExt,
+	plonk::{ConstraintSystem, Constraints, Error, Selector},
+	poly::Rotation,
+};
+
+use super::lt_eq_lookup::{MockChip, MockChipset, MockCommonConfig};
+
+/// Short length bits check chip using lookup table.
+///
+/// Checks if the target value is `S` bits long.
+#[derive(Debug, Clone)]
+pub struct LookupShortWordCheckChip<F: FieldExt, const K: usize, const S: usize> {
+	x: AssignedCell<F, F>,
+}
+
+impl<F: FieldExt, const K: usize, const S: usize> LookupShortWordCheckChip<F, K, S> {
+	/// Construct new instance
+	pub fn new(x: AssignedCell<F, F>) -> Self {
+		Self { x }
+	}
+}
+
+impl<F: FieldExt, const K: usize, const S: usize> MockChip<F>
+	for LookupShortWordCheckChip<F, K, S>
+{
+	type Output = AssignedCell<F, F>;
+
+	fn configure(mock_common: &MockCommonConfig, meta: &mut ConstraintSystem<F>) -> Selector {
+		assert!(0 < S && S < K, "Word bits should be less than target bits.");
+
+		let bitshift_selector = meta.complex_selector();
+
+		let word_column = mock_common.common.advice[0];
+
+		meta.lookup("Check K bits limit", |meta| {
+			let bitshift_selector = meta.query_selector(bitshift_selector);
+			let shifted_word = meta.query_advice(word_column, Rotation::cur());
+
+			vec![(bitshift_selector * shifted_word, mock_common.table)]
+		});
+
+		// For short lookups, check that the word has been shifted by the correct number
+		// of bits. https://p.z.cash/halo2-0.1:decompose-short-lookup
+		meta.create_gate("Short word S bits limit", |meta| {
+			let bitshift_selector = meta.query_selector(bitshift_selector);
+			let word = meta.query_advice(word_column, Rotation::prev());
+			let shifted_word = meta.query_advice(word_column, Rotation::cur());
+			let inv_two_pow_s = meta.query_advice(word_column, Rotation::next());
```

Comment:
Why use 1 column and multiple rows, when there are available columns left unused.
The general idea when designing constraints is to use up all the columns in one row until there are none left, then move to the next row using rotations

Reply:
For `LookupRangeCheckChip` this is hard to do, since we are calculating running sum for array of words (and array can have variable length). We would have to split the array into chunks, when the length of the chunk is the same as the amount of columns.
This will be hard to do, so we can leave it for the future.
But `ShortWordCheckChip` is very easy since we just need 2 or 3 columns

Reply:
Yes, the `ShortWordCheckChip` needs just 2 columns if we hard-code the `S` value like `K`.

Reply:
Done https://github.com/eigen-trust/protocol/pull/169/commits/a5db03110ec94caff1837ee9d9dda57ccfbac444